In [109]:
# import kagglehub

# # Download latest version
# path = kagglehub.competition_download('spaceship-titanic',output_dir='./data/')

# print("Path to competition files:", path)

In [110]:
import pandas as pd

In [111]:
df_train = pd.read_csv('./data/train.csv')
df_test = pd.read_csv('./data/test.csv')

In [112]:
df_train.shape

(8693, 14)

In [113]:
df_test.shape

(4277, 13)

In [114]:
df_train.head()

,PassengerId,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Name,Transported
0,0001_01,Europa,False,B/0/P,TRAPPIST-1e,39.0,False,0.0,0.0,0.0,0.0,0.0,Maham Ofracculy,False
1,0002_01,Earth,False,F/0/S,TRAPPIST-1e,24.0,False,109.0,9.0,25.0,549.0,44.0,Juanna Vines,True
2,0003_01,Europa,False,A/0/S,TRAPPIST-1e,58.0,True,43.0,3576.0,0.0,6715.0,49.0,Altark Susent,False
3,0003_02,Europa,False,A/0/S,TRAPPIST-1e,33.0,False,0.0,1283.0,371.0,3329.0,193.0,Solam Susent,False
4,0004_01,Earth,False,F/1/S,TRAPPIST-1e,16.0,False,303.0,70.0,151.0,565.0,2.0,Willy Santantines,True


In [115]:
df_test.head()

,PassengerId,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Name
0,0013_01,Earth,True,G/3/S,TRAPPIST-1e,27.0,False,0.0,0.0,0.0,0.0,0.0,Nelly Carsoning
1,0018_01,Earth,False,F/4/S,TRAPPIST-1e,19.0,False,0.0,9.0,0.0,2823.0,0.0,Lerome Peckers
2,0019_01,Europa,True,C/0/S,55 Cancri e,31.0,False,0.0,0.0,0.0,0.0,0.0,Sabih Unhearfus
3,0021_01,Europa,False,C/1/S,TRAPPIST-1e,38.0,False,0.0,6652.0,0.0,181.0,585.0,Meratz Caltilter
4,0023_01,Earth,False,F/5/S,TRAPPIST-1e,20.0,False,10.0,0.0,635.0,0.0,0.0,Brence Harperez


In [116]:
df_train.columns

Index(['PassengerId', 'HomePlanet', 'CryoSleep', 'Cabin', 'Destination', 'Age',
       'VIP', 'RoomService', 'FoodCourt', 'ShoppingMall', 'Spa', 'VRDeck',
       'Name', 'Transported'],
      dtype='str')

In [117]:
df_train.columns[df_train.dtypes != 'float64']

Index(['PassengerId', 'HomePlanet', 'CryoSleep', 'Cabin', 'Destination', 'VIP',
       'Name', 'Transported'],
      dtype='str')

In [118]:
c_train=df_train.drop(columns=['PassengerId'])

In [119]:
c_test=df_test.copy()

### c_test=df_test.drop(columns=['PassengerId'])
#### drping it would make it unrecoverable in the submission file, so we will keep it for now

In [120]:
c_train.columns[c_train.dtypes != 'float64']

Index(['HomePlanet', 'CryoSleep', 'Cabin', 'Destination', 'VIP', 'Name',
       'Transported'],
      dtype='str')

In [121]:
missing_per_row = c_train.isnull().sum(axis=1)

print("Rows with 0 missing:", (missing_per_row == 0).sum())
print("Rows with 1 missing:", (missing_per_row == 1).sum())
print("Rows with 2 missing:", (missing_per_row == 2).sum())
print("Rows with 3 missing:", (missing_per_row == 3).sum())
print("Rows with 4 missing:", (missing_per_row == 4).sum())
print("Rows with 5 missing:", (missing_per_row == 5).sum())
print("Rows with 6 missing:", (missing_per_row == 6).sum())
print("Rows with ≥7 missing:", (missing_per_row >= 7).sum())

Rows with 0 missing: 6606
Rows with 1 missing: 1867
Rows with 2 missing: 203
Rows with 3 missing: 17
Rows with 4 missing: 0
Rows with 5 missing: 0
Rows with 6 missing: 0
Rows with ≥7 missing: 0


In [122]:
c_train.isnull().sum()

HomePlanet      201
CryoSleep       217
Cabin           199
Destination     182
Age             179
VIP             203
RoomService     181
FoodCourt       183
ShoppingMall    208
Spa             183
VRDeck          188
Name            200
Transported       0
dtype: int64

In [123]:
# from sklearn.model_selection import train_test_split
# from sklearn.ensemble import RandomForestClassifier
# from sklearn.metrics import accuracy_score

# # Use only non-null rows for Transported
# train_df = df[df['Transported'].notnull()]

# # Features: categorical predictors
# X = pd.get_dummies(train_df[['HomePlanet','CryoSleep','Destination','VIP']], drop_first=True)
# y = train_df['Transported']

# # Split into train/test
# X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# # Train model
# model = RandomForestClassifier(random_state=42)
# model.fit(X_train, y_train)

# # Evaluate
# y_pred = model.predict(X_test)
# print("We are testing tht can we replace the missing values with the preddicted values by the model and check the accuracy of the model")
# print("Accuracy:", accuracy_score(y_test, y_pred))


#### We are testing tht can we replace the missing values with the preddicted values by the model and check the accuracy of the model. Accuracy: 0.7228292121909143

### not dropping in test data because we need to predict the missing values in test data

In [124]:
print("Missing values in test data (%):")
print(c_test.isna().sum()/c_test.shape[0]*100)

Missing values in test data (%):
PassengerId     0.000000
HomePlanet      2.034136
CryoSleep       2.174421
Cabin           2.338087
Destination     2.151040
Age             2.127660
VIP             2.174421
RoomService     1.917232
FoodCourt       2.478373
ShoppingMall    2.291326
Spa             2.361468
VRDeck          1.870470
Name            2.197802
dtype: float64


In [125]:
for col in ['HomePlanet','CryoSleep','Destination','VIP']:
    c_test[col] = c_test[col].fillna("Unknown")
for col in ['HomePlanet','CryoSleep','Destination','VIP']:
    c_train[col] = c_train[col].fillna("Unknown")
c_train=c_train.dropna()
c_train.isnull().sum()

HomePlanet      0
CryoSleep       0
Cabin           0
Destination     0
Age             0
VIP             0
RoomService     0
FoodCourt       0
ShoppingMall    0
Spa             0
VRDeck          0
Name            0
Transported     0
dtype: int64

In [126]:
print(c_train['HomePlanet'].value_counts())
print(c_train['CryoSleep'].value_counts())
print(c_train['Destination'].value_counts())
print(c_train['VIP'].value_counts())
print(c_train['Transported'].value_counts())
print(f"nunique name: {c_train['Name'].nunique()}")
print(f"nunique cabin: {c_train['Cabin'].nunique()}")

HomePlanet
Earth      3834
Europa     1789
Mars       1471
Unknown     171
Name: count, dtype: int64
CryoSleep
False      4597
True       2492
Unknown     176
Name: count, dtype: int64
Destination
TRAPPIST-1e      4936
55 Cancri e      1504
PSO J318.5-22     670
Unknown           155
Name: count, dtype: int64
VIP
False      6918
Unknown     176
True        171
Name: count, dtype: int64
Transported
True     3655
False    3610
Name: count, dtype: int64
nunique name: 7249
nunique cabin: 5743


In [127]:
c_train=c_train.drop(columns=['Cabin','Name'])

In [128]:
c_test=c_test.drop(columns=['Cabin','Name'])

In [129]:
print("Missing values in test data (%):")
print(c_test.isna().sum()/c_test.shape[0]*100)

Missing values in test data (%):
PassengerId     0.000000
HomePlanet      0.000000
CryoSleep       0.000000
Destination     0.000000
Age             2.127660
VIP             0.000000
RoomService     1.917232
FoodCourt       2.478373
ShoppingMall    2.291326
Spa             2.361468
VRDeck          1.870470
dtype: float64


In [130]:
# Fill numeric NaNs with median directly
num_cols = ['Age','RoomService','FoodCourt','ShoppingMall','Spa','VRDeck']
c_test[num_cols] = c_test[num_cols].fillna(c_test[num_cols].median())

In [131]:
print("Missing values in test data (%):")
print(c_test.isna().sum()/c_test.shape[0]*100)

Missing values in test data (%):
PassengerId     0.0
HomePlanet      0.0
CryoSleep       0.0
Destination     0.0
Age             0.0
VIP             0.0
RoomService     0.0
FoodCourt       0.0
ShoppingMall    0.0
Spa             0.0
VRDeck          0.0
dtype: float64


## 🧹 Missing Data Handling

* **Categorical:** Filled missing values with `"Unknown"`.
* **Numeric:** Filled missing values with the median.
* **`Cabin`:** Dropped because cabin details were shared by some passengers and offered limited individual information.
* **`Name`:** Dropped because names were mostly unique and not useful for generalization.


In [132]:
c_train=pd.get_dummies(c_train, columns=['HomePlanet','CryoSleep','Destination','VIP',], drop_first=True,dtype=float)

# Transported is target data so it is not avaible in test data

In [133]:
c_test=pd.get_dummies(c_test, columns=['HomePlanet','CryoSleep','Destination','VIP',], drop_first=True,dtype=float)

In [134]:
c_train.columns[c_train.dtypes != 'float64']

Index(['Transported'], dtype='str')

In [135]:
c_train.shape[1]==c_test.shape[1]+1

False

In [136]:
c_test.shape[0]==4277

True

In [137]:
pd.to_pickle(c_train, './cdata/c_train.pkl')
pd.to_pickle(c_test, './cdata/c_test.pkl')
c_train.to_csv('./cdata/c_train.csv', index=False)
c_test.to_csv('./cdata/c_test.csv', index=False)

## 🧹 Dataset Preprocessing — Summary

* Loaded **8,693 training** and **4,277 test** passengers.
* Removed `PassengerId` as it is only an identifier.
* Investigated missing values; most rows had **0–1 missing values** and each feature had only ~2% missing data.
* Tested **model-based imputation** for categorical values, but accuracy was only ~72%, so this approach was not used.
* Filled missing categorical values (`HomePlanet`, `CryoSleep`, `Destination`, `VIP`) with `"Unknown"`.
* Dropped `Cabin` and `Name`: `Cabin` provided limited individual information, while `Name` was mostly unique.
* Filled missing numeric values (`Age`, `RoomService`, `FoodCourt`, `ShoppingMall`, `Spa`, `VRDeck`) with the **median**.
* Converted categorical features to **numeric dummy variables** using one-hot encoding.
* Verified that train/test feature counts matched correctly and that **no missing values remained**.
* Saved the processed datasets as `.csv` and `.pkl` files in `cdata/`.

### 🔎 Key Experiment

Model-based filling of categorical missing values was tested using a Random Forest, but the validation accuracy was only **~72.3%**. Therefore, `"Unknown"` was used instead, preserving the fact that the original value was missing.
